csaf/src/metadata_retriever/mod.rs

```diff
@@ -0,0 +1,200 @@
+                                Ok(None)
+                            } else {
+                                Err(Error::Fetcher(fetcher::Error::Request(reqwest_error)))
+                            }
+                        }
+                    },
+                    _ => Err(err),
+                },
+            }
+        } else {
+            Err(Error::Url(ParseError::EmptyHost))
+        }
+    }
+
+    /// Retrieve provider metadata through the security text of provided URL.
+    pub async fn retrieve_metadata_by_security_text(
+        &self,
+        provider_metadata_url: Url,
+        security_txt_path: &str,
+    ) -> Result<Option<Json<ProviderMetadata>>, Error> {
+        log::trace!(
+            "Starting retrieve provider metadata from security text of provided discovery URL "
+        );
+        if let Some(host_url) = provider_metadata_url.clone().host_str() {
+            let host_url_string = format!("https://{}", &host_url);
+            let host_url = Url::parse(host_url_string.as_str())?;
+            log::info!(
+                "Querying provider metadata url from security text of {:?}",
+                host_url_string
+            );
+            if let Some(path) = Self::get_metadata_url_from_security_text(
+                &self.fetcher,
+                host_url.join(security_txt_path)?,
+            )
+            .await?
+            {
+                let provider_metadata_url = Url::parse(path.as_str())?;
+
+                Ok(self.fetch_metadata_from_url(provider_metadata_url).await?)
+            } else {
+                log::warn!(
+                    "The Security Text obtained from this URL is 'None': {} ",
+                    host_url_string
+                );
+                Ok(None)
+            }
+        } else {
+            Err(Error::Url(ParseError::EmptyHost))
+        }
+    }
+
+    ///For the provided metadata URL, retrieve the provider metadata using three different discovery methods in a specified order.
+    pub async fn retrieve(&self) -> Result<DetectionType, Error> {
+        let actually_url = if self.base_url.clone().starts_with("https://") {
+            Url::parse(&self.base_url)?
+        } else {
+            Url::parse(&format!("https://{}", self.base_url.clone()))?
+        };
+
+        if let Some(provider_metadata) = self
+            .retrieve_metadata_by_well_known_url(actually_url.clone())
+            .await?
+        {
+            return Ok(DetectionType::WellKnowPath(provider_metadata.into_inner()));
+        } else {
+            log::warn!(
+                    "The provider metadata obtained from this well known URL is 'None': {}，the process will proceed to the next discovery rule.",
+                    self.base_url.clone().to_string(),
```

Comment:
`self.url` implements `Display`, no need to either clone nor call to_string. Same for the other locations.